### PR TITLE
Rework SSE passive loop

### PR DESCRIPTION
**Pull request by achouippe**
This PR bring somes reworks:

**Graceful shutdown is now message based**
- SSE processes register themself to a process [Registry](https://hexdocs.pm/elixir/1.12/Registry.html),
- When the mix application is asked to stop, all process registered in the registry receives a `shutdown` message,
- SSE processes react to this message by sending a `shutdown` SSE event and then stop the HTTP connection immediately,

**Dynamic wait delay in the SSE connection loop**

This changes allow to rework the passive loop of SSE connections: The passive loop does not rely anymore on a 1s fixed wait between each loop. The duration between two loops is now computed from the duration expected before than:
- A keep alive must be sent to the client,
- Or the connection will timeout,
- Or the JWT token will expire.

This PR should slightly reduce the CPU consumption by reducing the amounts of "loops" per SSE connection.

### CODE REVIEW

File: neurow/lib/neurow/application.ex [hunk 1]
```diff
@@ -74,8 +74,8 @@ defmodule Neurow.Application do
          scheme: :http, plug: Neurow.InternalApi.Endpoint, options: [port: internal_api_port]},
         {Plug.Cowboy,
          scheme: sse_http_scheme, plug: Neurow.PublicApi.Endpoint, options: public_api_http_config},
-        {Plug.Cowboy.Drainer, refs: [Neurow.PublicApi.Endpoint], shutdown: 20_000},
-        {StopListener, []},
+        {Plug.Cowboy.Drainer, refs: [Neurow.PublicApi.Endpoint.HTTP], shutdown: 20_000},
```
Comment (achouippe) [inline neurow/lib/neurow/application.ex:77]: Discovered a regression introduced in a previous PR, prefixing by `.HTTP` fixes it.

File: neurow/lib/neurow/public_api/endpoint.ex [hunk 2]
```diff
@@ -194,58 +196,84 @@ defmodule Neurow.PublicApi.Endpoint do
...
-          now_ms - last_message > sse_timeout ->
-            Logger.debug("Client disconnected due to inactivity")
-            conn |> write_raw_chunk("event: timeout")
-
-          # SSE Keep alive, send a ping
-          now_ms - last_ping > keep_alive ->
-            conn
-            |> write_raw_chunk("event: ping")
-            |> loop(sse_timeout, keep_alive, last_message, now_ms, jwt_exp)
+    cond do
```
Comment (achouippe) [inline neurow/lib/neurow/public_api/endpoint.ex:202]: I reversed the order of the `receive` block and all the conditions in the `cond` block in the last commit.

If it less readable this way, I can just revert the last commit on the branch.

File: neurow/lib/neurow/public_api/endpoint.ex [hunk 3]
```diff
@@ -194,58 +196,84 @@ defmodule Neurow.PublicApi.Endpoint do
...
+        |> loop(
+          sse_timeout_ms,
+          keep_alive_ms,
+          last_message_ts,
+          now_ms,
+          jwt_exp
+        )
+
+      # JWT token expired
+      jwt_exp * 1000 < now_ms ->
```
Comment (bpaquet) [inline neurow/lib/neurow/public_api/endpoint.ex:221]:
```suggestion
      jwt_exp_s * 1000 < now_ms ->
```

File: neurow/lib/neurow/public_api/endpoint.ex [hunk 4]
```diff
@@ -194,58 +196,84 @@ defmodule Neurow.PublicApi.Endpoint do
...
+              new_last_message_ts,
+              jwt_exp
+            )
 
-          # We need to stop
-          StopListener.close_connections?() ->
-            conn |> write_raw_chunk("event: reconnect")
+          {:shutdown, _value} ->
+            Logger.debug("Client on disconnected due to node shutdown")
+            conn |> write_chunk("event: node_shutdown")
```
Comment (bpaquet) [inline neurow/lib/neurow/public_api/endpoint.ex:252]: Can we use a terminology for client?

Comment (achouippe) [reply neurow/lib/neurow/public_api/endpoint.ex:252]: What do you mean ? You prefer "reconnect" ?

File: neurow/lib/stop_listener.ex [hunk 5]
```diff
@@ -1,30 +1,40 @@
...
   def start_link(opts) do
     GenServer.start_link(__MODULE__, opts, name: __MODULE__)
   end
 
-  @impl true
-  def init(_) do
-    :ets.new(__MODULE__, [:set, :named_table, read_concurrency: true])
-    Process.flag(:trap_exit, true)
-    {:ok, %{shutdown_in_progress: false}}
+  def subscribe(value \\ nil) do
```
[the remark this answers is not in the document]
Comment (achouippe) [reply neurow/lib/stop_listener.ex:9]: It is possible to pass additional data while registering a process, I wanted initially to keep this possibility. But yes I can remove it.

File: neurow/lib/neurow/public_api/endpoint.ex (hunk 2 above)
Comment (papey) [inline neurow/lib/neurow/public_api/endpoint.ex:202]:
```

```

File: neurow/lib/stop_listener.ex [hunk 6]
```diff
@@ -1,30 +1,40 @@
...
+    case Registry.register(Registry.StopListener, :shutdown_subscribers, value) do
+      {:ok, _pid} -> :ok
```
Comment (papey) [inline neurow/lib/stop_listener.ex:11]: **Suggestion** 🤔 : I think it's ok to keep the tuple as is

File: neurow/lib/stop_listener.ex [hunk 7]
```diff
@@ -1,30 +1,40 @@
...
-      ArgumentError -> true
-    end
+  def shutdown() do
+    GenServer.cast(__MODULE__, :shutdown)
+  end
+
+  @impl true
+  def init(_opts) do
+    Registry.start_link(keys: :duplicate, name: Registry.StopListener)
+    {:ok, %{}}
```
Comment (papey) [inline neurow/lib/stop_listener.ex:23]: **Question** ❓ : is it a hard requirement to send this empty map ?

**Review by papey — COMMENTED** [review]
Looks good overall, small comment on readability, and I think distinct terms for server (shutdown) vs client (reconnect) will improve understanding overall

File: neurow/lib/neurow/public_api/endpoint.ex [hunk 8]
```diff
@@ -73,13 +74,14 @@ defmodule Neurow.PublicApi.Endpoint do
...
             Logger.debug("Client subscribed to #{topic}")
 
-            last_message = :os.system_time(:millisecond)
-            conn |> loop(timeout, keep_alive, last_message, last_message, exp)
+            now_ms = :os.system_time(:millisecond)
+
+            conn
+            |> send_chunked(200)
+            |> import_history(topic, last_event_id)
+            |> loop(timeout_ms, keep_alive_ms, now_ms, now_ms, exp)
```
[the remark this answers is not in the document]
Comment (achouippe) [reply neurow/lib/neurow/public_api/endpoint.ex:84]: Depending on the situation we use now_ms both as a timestamp for the last_message and last_ping.

File: neurow/lib/neurow/public_api/endpoint.ex [hunk 9]
```diff
@@ -37,13 +37,13 @@ defmodule Neurow.PublicApi.Endpoint do
       %{"iss" => issuer, "sub" => sub, "exp" => exp} ->
         topic = "#{issuer}-#{sub}"
 
-        timeout =
+        timeout_ms =
```
[the remark this answers is not in the document]
Comment (achouippe) [reply neurow/lib/neurow/public_api/endpoint.ex:40]: After discussing with @bpaquet , we do not need anymore to accept a x-sse-timeout header. I will include the removal in the current PR.

File: neurow/integration_test/sse_livecycle_test.exs [hunk 10]
```diff
@@ -102,9 +69,59 @@ defmodule Neurow.IntegrationTest.SseLifecycleTest do
...
+      {node_name, public_api_port, _internal_api_port} = TestCluster.start_new_node()
+
+      on_exit(fn ->
+        TestCluster.shutdown_node(node_name)
+      end)
+
+      {:ok, node_name: node_name, public_api_port: public_api_port}
+    end
+
+    test "the client receives a 'shutdown' event when the node it is connected to is shutdowned",
```
Comment (papey) [inline neurow/integration_test/sse_livecycle_test.exs:88]: **Question** ❓ : is the final wording shutdown of reconnect ?

**Review by papey — APPROVED** [review]

**Pull request merged**